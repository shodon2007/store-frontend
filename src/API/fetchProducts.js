import axios from "axios";

export async function getProducts(type) {
    const resp = await axios.get(`http://localhost:3000/products/${type}`);
    return resp.data;
}

export async function getProduct(type, id) {
    const resp = await axios.get(`http://localhost:3000/products/${type}/${id}`);
    return resp.data;
}