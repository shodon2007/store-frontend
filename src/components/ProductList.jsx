import React from 'react'
import { Link, useParams } from 'react-router-dom'
import { URL } from '../consts/consts'
import { useProducts } from '../hooks/useProducts';

const ProductList = ({ brand }) => {
    const { type } = useParams();
    const { isFetching, data } = useProducts(brand, type);

    if (isFetching) {
        return <div>Загрузка данных...</div>
    }

    if (data.length === 0 || !data) {
        return <h1>404, Ничего не найдено, сорри чувак =((</h1>
    }

    return (
        <div>
            <div className="products__list">
                {data.map(product => {
                    return <Link
                        key={product.name}
                        className='products__product'
                        to={`${product.id}`}
                    >
                        <img src={`${URL}/${product.img}`} alt="product-img" />
                        {product.name}
                    </Link>
                })}
            </div>
        </div>
    )
}

export default ProductList