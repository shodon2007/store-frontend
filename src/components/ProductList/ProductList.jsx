import React from 'react'
import { Link, useParams } from 'react-router-dom'

import { URL } from '../../consts/consts'
import { useProducts } from '../../hooks/useProducts';
import Error404 from '../../pages/Error404';
import MyText from '../UI/text/MyText';
import MyPrice from '../UI/price/MyPrice';

import classes from './ProductList.module.scss';
import Loading from '../../pages/Loading';

const ProductList = ({ brand }) => {
    const { type } = useParams();
    const { isFetching, data } = useProducts(brand, type);

    if (isFetching) {
        return <Loading />
    }

    const notFound = data.length === 0 || !data;

    if (notFound) {
        return <Error404 />
    }

    return (
        <div className={classes.list}>
            {data.map(product => {
                return (
                    <Link
                        key={product.name}
                        className={classes.product}
                        to={`${product.id}`}
                    >
                        <img className={classes.img} src={`${URL}/${product.img}`} alt="product-img" />
                        <div className={classes.bottom}>
                            <MyText>{product.name}</MyText>
                            <MyPrice>{product.price} рублей</MyPrice>
                        </div>
                    </Link>
                )
            })}
        </div>
    )
}

export default ProductList