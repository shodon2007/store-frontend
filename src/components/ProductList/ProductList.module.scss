@import "../../styles/vars.scss";

.list {
    display: grid;
    grid-template-columns: repeat(1, 1fr);
    background-color: white;
}

.item {
    display: grid;
    grid-template-columns: 250px 1fr max-content;
    padding: 20px 30px;
    border-radius: 10px;
    gap: 10px;
    font-size: 16px;
    color: black;
    text-decoration: none;
    height: min-content;
    transition: all 0.3s;
}

.item:hover * {
    transition: all 0.3s;
    color: $bgAll;
}

.img {
    max-width: 100%;
    min-height: 120px;
    min-width: 200px;
    background: #b8b8b8;
}

.bottom {
    display: flex;
    flex-direction: column;
    flex: 1 1 auto;
}

@media (max-width: 1200px) {
    .list {
        gap: 10px;
    }
    .item {
        padding: 5px 10px;
    }
}

@media (max-width: 700px) {
    .list {
        grid-template-columns: repeat(2, 1fr);
    }
    .item {
        height: 100%;
        grid-template-columns: 1fr;
        grid-template-rows: 1fr max-content max-content;
        gap: 5px;
    }
}
