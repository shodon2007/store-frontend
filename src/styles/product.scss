.product {
    display: flex;
    flex-direction: column;
    gap: 10px;
}

.product__item {
    display: grid;
    grid-template-columns: max-content 1fr;
    gap: 10px;

    &>img {
        max-width: 100%;
    }
}

.attribute__item {
    display: flex;
}

.product__attributes {
    display: flex;
    flex-direction: column;
    gap: 10px;
}

.product__bottom {
    display: flex;
    justify-content: space-between;
}

.product__price {
    font-size: 20px;
    font-weight: bold;
}

.product__basket {
    font-size: 25px;
    cursor: pointer;
    padding: 10px 20px;
    background-color:  #10C44C;
    border: 0;
    border-radius: 10px;
}

.product__info {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
}