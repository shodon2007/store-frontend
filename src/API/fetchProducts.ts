import { TypeGetFilter, TypeGetProduct, TypeGetProducts } from "../types/product";
import defaultAxios from "./axiosService";

export const getProducts: TypeGetProducts = async (type, brand, form) => {
    return await defaultAxios.get(`/products/${type}`, {
        params: {
            brand,
            form
        }
    });
}

export const getProduct: TypeGetProduct = async (type, id) => {
    return await defaultAxios.get(`/products/${type}/${id}`);
}

export const getFilter: TypeGetFilter = async (type) => {
    return await defaultAxios.get(`/side/${type}`);
}