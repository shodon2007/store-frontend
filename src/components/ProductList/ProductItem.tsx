import { FC, memo } from "react";
import { IProduct } from "../../types/product";
import { Link } from "react-router-dom";

type Props = {
    item: IProduct;
};

import classes from "./ProductList.module.scss";
import MyText from "../UI/text/MyText";
import MyPrice from "../UI/price/MyPrice";
import { URL } from "../../consts/consts";

const ProductItem: FC<Props> = memo(({ item }) => {
    return (
        <Link key={item.id} className={classes.item} to={`${item.id}`}>
            <img
                className={classes.img}
                src={`${URL}/${item.img}`}
                alt="product-img"
            />
            <div className={classes.bottom}>
                <MyText>{item.name}</MyText>
                <MyPrice>{item.price} рублей</MyPrice>
            </div>
        </Link>
    );
});

export default ProductItem;
