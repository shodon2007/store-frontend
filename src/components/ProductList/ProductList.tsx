import classes from "./ProductList.module.scss";
import { FC, memo } from "react";
import { IProduct } from "../../types/product";
import ProductItem from "./ProductItem";
import Error404 from "../../pages/Error404";

type TypeList = {
    itemList: IProduct[];
    page: number;
};

const ProductList: FC<TypeList> = memo(({ itemList, page }) => {
    if (itemList.length === 0) {
        return <Error404 />;
    }

    const pageItems = page * 5;

    itemList = itemList.slice(pageItems, pageItems + 5);

    return (
        <div className={classes.list}>
            {itemList.map((item: IProduct) => {
                return <ProductItem item={item} />;
            })}
        </div>
    );
});

export default ProductList;
