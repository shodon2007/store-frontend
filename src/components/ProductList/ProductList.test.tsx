import { render, screen } from "@testing-library/react";
import ProductList from "./ProductList";
import { IProduct } from "../../types/product";
import { MemoryRouter } from "react-router-dom";
import { getProducts } from "../../API/fetchProducts";
import { TypeForm } from "../../types/side";

const form: TypeForm = {
    filter: {},
    brands: [],
    price: { min: 0, max: 1000000 },
    sort: "name",
};

describe("Testing Product List", () => {
    test("Testing empty List", () => {
        const itemList: IProduct[] = [];
        render(<ProductList itemList={itemList} />);
        const emptyTitle = screen.getByRole("heading");
        expect(emptyTitle).not.toBeNull();
        expect(emptyTitle).toMatchSnapshot();
    });
    test("Testing item phones", async () => {
        const { data } = await getProducts("phones", form);

        render(
            <MemoryRouter>
                <ProductList itemList={data} />
            </MemoryRouter>
        );

        screen.debug();

        const links = screen.getAllByRole("link");
        expect(links.length).toBe(data.length);
        links.forEach((item) => {
            expect(item).not.toBeNull();
            expect(item).toMatchSnapshot();
        });
    });
});
