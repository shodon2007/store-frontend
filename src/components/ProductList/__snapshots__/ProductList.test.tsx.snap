// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Product List Testing empty List 1`] = `
<h1
  class="title"
>
  404, Ничего не найдено
</h1>
`;

exports[`Testing Product List Testing item phones 1`] = `
<a
  class="item"
  href="/13"
>
  <img
    alt="product-img"
    class="img"
    src="https://15f3e04d9857.vps.myjino.ru//google6.webp"
  />
  <div
    class="text"
  >
    Google Pixel 6
  </div>
  <div
    class="price"
  >
    50000
     ₽
  </div>
</a>
`;

exports[`Testing Product List Testing item phones 2`] = `
<a
  class="item"
  href="/12"
>
  <img
    alt="product-img"
    class="img"
    src="https://15f3e04d9857.vps.myjino.ru//google7.webp"
  />
  <div
    class="text"
  >
    Google Pixel 7
  </div>
  <div
    class="price"
  >
    60000
     ₽
  </div>
</a>
`;

exports[`Testing Product List Testing item phones 3`] = `
<a
  class="item"
  href="/7"
>
  <img
    alt="product-img"
    class="img"
    src="https://15f3e04d9857.vps.myjino.ru//huaweip60.webp"
  />
  <div
    class="text"
  >
    Huawei p60 pro
  </div>
  <div
    class="price"
  >
    100000
     ₽
  </div>
</a>
`;

exports[`Testing Product List Testing item phones 4`] = `
<a
  class="item"
  href="/4"
>
  <img
    alt="product-img"
    class="img"
    src="https://15f3e04d9857.vps.myjino.ru//iphone12.webp"
  />
  <div
    class="text"
  >
    Iphone 12 pro max
  </div>
  <div
    class="price"
  >
    150000
     ₽
  </div>
</a>
`;

exports[`Testing Product List Testing item phones 5`] = `
<a
  class="item"
  href="/1"
>
  <img
    alt="product-img"
    class="img"
    src="https://15f3e04d9857.vps.myjino.ru//iphone13promax.webp"
  />
  <div
    class="text"
  >
    Iphone 13 pro max
  </div>
  <div
    class="price"
  >
    180000
     ₽
  </div>
</a>
`;

exports[`Testing Product List Testing item phones 6`] = `
<a
  class="item"
  href="/3"
>
  <img
    alt="product-img"
    class="img"
    src="https://15f3e04d9857.vps.myjino.ru//iphone14.webp"
  />
  <div
    class="text"
  >
    Iphone 14 pro max
  </div>
  <div
    class="price"
  >
    200000
     ₽
  </div>
</a>
`;

exports[`Testing Product List Testing item phones 7`] = `
<a
  class="item"
  href="/11"
>
  <img
    alt="product-img"
    class="img"
    src="https://15f3e04d9857.vps.myjino.ru//pocox3.webp"
  />
  <div
    class="text"
  >
    Poco x3
  </div>
  <div
    class="price"
  >
    40000
     ₽
  </div>
</a>
`;

exports[`Testing Product List Testing item phones 8`] = `
<a
  class="item"
  href="/10"
>
  <img
    alt="product-img"
    class="img"
    src="https://15f3e04d9857.vps.myjino.ru//pocox4.webp"
  />
  <div
    class="text"
  >
    Poco x4
  </div>
  <div
    class="price"
  >
    60000
     ₽
  </div>
</a>
`;

exports[`Testing Product List Testing item phones 9`] = `
<a
  class="item"
  href="/9"
>
  <img
    alt="product-img"
    class="img"
    src="https://15f3e04d9857.vps.myjino.ru//pocox5.webp"
  />
  <div
    class="text"
  >
    Poco x5
  </div>
  <div
    class="price"
  >
    80000
     ₽
  </div>
</a>
`;

exports[`Testing Product List Testing item phones 10`] = `
<a
  class="item"
  href="/6"
>
  <img
    alt="product-img"
    class="img"
    src="https://15f3e04d9857.vps.myjino.ru//samsunga54.webp"
  />
  <div
    class="text"
  >
    Samsung a54
  </div>
  <div
    class="price"
  >
    36000
     ₽
  </div>
</a>
`;

exports[`Testing Product List Testing item phones 11`] = `
<a
  class="item"
  href="/2"
>
  <img
    alt="product-img"
    class="img"
    src="https://15f3e04d9857.vps.myjino.ru//samsungs21.webp"
  />
  <div
    class="text"
  >
    Samsung Galaxy S21 FE
  </div>
  <div
    class="price"
  >
    54000
     ₽
  </div>
</a>
`;

exports[`Testing Product List Testing item phones 12`] = `
<a
  class="item"
  href="/5"
>
  <img
    alt="product-img"
    class="img"
    src="https://15f3e04d9857.vps.myjino.ru//samsungs23.webp"
  />
  <div
    class="text"
  >
    Samsung s23 ultra
  </div>
  <div
    class="price"
  >
    70000
     ₽
  </div>
</a>
`;
